### ISSUE TITLE
Issues with selecting items from this widget

### ISSUE THREAD

**Issue by jking6884** [toranb/ember-cli-auto-complete#7]
When you click the input, the suggestions drop down and you can click one.  However, if you click on the input and select a different options several times, it seems to eventually have issues firing the selectItem action.  One this happens, you can double click the items in the drop down list and it will select it.  You can also change the text in the text box by typing and then clicking something from the suggestions will work as well.  However, if we take your jsbin example for instance, if you use this widget like a select box (as in, you just click the box and then click an option), it will eventually get hung up and you will not be able to get the widget to update the selection by clicking on one of the options without doing one of the above mentioned things.  Another issue is that, again using the jsbin example, if you click the widget, and when the suggestions appear, you click ABCDE, when you click the widget again, all of the options appear in the suggestions.  However, since ABCDE has been selected, all of the other options should not be there.  It seems that if you use this widget strictly by typing text and then using the keyboard to move up and down and using the enter key to select an option, it works 100% of the time.  However, there seem to be issues with click events when using it in the ways described above.

**Comment by toranb** [comment]
Wow thanks for the rocking feedback! I need to verify I can reproduce this w/ pure ember-testing first and if not (because your results are very mouse driven) I might need to result to selenium (hope not but it's an option).

Thanks for taking the time to write this up - I'll get w/ @abuiles to see what we can do to better ensure click events are sent/bubbld' up :)

**Comment by toranb** [comment]
Also - just to clarify  ~ are you using ember 1.11 (to help me reproduce this)

**Comment by jking6884** [comment]
Thanks for getting back to me as well!  Btw, we are using a couple of your widgets in our apps now with great success and are wanting to implement auto complete as well.  Really appreciate the work you guys are doing!

```
 On Monday, March 30, 2015 9:00 PM, Toran Billups <notifications@github.com> wrote:
```

 Also - just to clarify ~ are you using ember 1.11 (to help me reproduce this)—

**Comment by jking6884** [comment]
Yes, I updated the project today to 1.11.

```
 On Monday, March 30, 2015 10:23 PM, John King <jking6884@yahoo.com> wrote:
```

 Thanks for getting back to me as well!  Btw, we are using a couple of your widgets in our apps now with great success and are wanting to implement auto complete as well.  Really appreciate the work you guys are doing!

```
 On Monday, March 30, 2015 9:00 PM, Toran Billups <notifications@github.com> wrote:
```

 Also - just to clarify ~ are you using ember 1.11 (to help me reproduce this)—

**Comment by jking6884** [comment]
Have you been able to 'see' the issue yet?  I can try to record a screen cast if it would be helpful? Don't know if my description was clear enough or not.  Just let me know.

```
 On Monday, March 30, 2015 10:27 PM, John King <jking6884@yahoo.com> wrote:
```

 Yes, I updated the project today to 1.11.

```
 On Monday, March 30, 2015 10:23 PM, John King <jking6884@yahoo.com> wrote:
```

 Thanks for getting back to me as well!  Btw, we are using a couple of your widgets in our apps now with great success and are wanting to implement auto complete as well.  Really appreciate the work you guys are doing!

```
 On Monday, March 30, 2015 9:00 PM, Toran Billups <notifications@github.com> wrote:
```

 Also - just to clarify ~ are you using ember 1.11 (to help me reproduce this)—

**Comment by toranb** [comment]
I was able to see the "double click required" behavior you mention. After playing around w/ the jsbin I can see how a combination of clicking + typing ahead and selecting/selecting again wasn't being picked up as you'd expect.

I still need to write a failing ember-testing acceptance-test or selenium test to prove it /fix it for real this week. I'll follow up later in the week when I get a free minute to hack this out :)

**Comment by toranb** [comment]
Just following up to confirm I didn't have time yet this week to write a failing test for this /look it over to fix the bug you found. Sorry for the delay - just working heavily on the identity map this week (doing a few 1.0 beta releases throughout the week).

**Comment by toranb** [comment]
A little more playing around and I landed a failing scenario (not a failing test just yet)

``` js
test("clicking the option directly will always set the value", function(assert) {
  visit("/");
  click("input.typeahead");
  click(".tt-suggestion:eq(2)");
  andThen(function() {
    assert.equal(find("input.typeahead").val(), "ABZ");
  });
  click(".tt-suggestion:eq(1)");
  andThen(function() {
    assert.equal(find("input.typeahead").val(), "ABD");
  });
  click(".tt-suggestion:eq(1)");
  andThen(function() {
    assert.equal(find("input.typeahead").val(), "ABD");
  });
  click(".tt-suggestion:eq(2)");
  andThen(function() {
    assert.equal(find("input.typeahead").val(), "ABZ");
  });
});
```

This is what I'm doing manually but it "passes" as a test currently. I'm hoping to learn why this acceptance test I wrote isn't _exactly_ like the way the browser behaves before I can dive in deeper to solve this issue.

**update**

My guess is that when I click around (quickly) in bowser this run later is playing a role that my acceptance tests might be ignoring ...

``` js
    Ember.run.later(this, func, 100);
```

**Comment by toranb** [comment]
Just to confirm - removing the focusOut block completely solves this problem. I'll dig in tonight to learn why :)

**Comment by toranb** [comment]
After a little hacking I also found the filter wasn't being reset :)

I just did a publish so look for 0.0.12 on npm. I also attempted to fix the clicking problem you reported - can you verify it's working better than it did previously? Any other issues let me know (jsbin was updated also to use the latest)

http://emberjs.jsbin.com/niluhaloke/1

**Comment by toranb** [comment]
I'll close this issue out for now but please reopen if you see any other click related problems / selection issues. Thanks again for reporting the bug !

**Comment by jking6884** [comment]
Thanks for following up.  I will definitely get the new code and check it out.

John

```
 On Thursday, April 9, 2015 11:10 PM, Toran Billups <notifications@github.com> wrote:
```

 I'll close this issue out for now but please reopen if you see any other click related problems / selection issues. Thanks again for reporting the bug !—

**Comment by jking6884** [comment]
Thanks for following up.  I will definitely get the new code and check it out.

John 

```
 On Thursday, April 9, 2015 11:36 PM, John King <jking6884@yahoo.com> wrote:
```

 Thanks for following up.  I will definitely get the new code and check it out.

John

```
 On Thursday, April 9, 2015 11:10 PM, Toran Billups <notifications@github.com> wrote:
```

 I'll close this issue out for now but please reopen if you see any other click related problems / selection issues. Thanks again for reporting the bug !—

**Comment by jking6884** [comment]
Toran,
I got your new code but it seems that the problem still persists.  The functionality works great if you are just using the keyboard, but if you type in "AB" (using your jsbin) and then try to click one of the options as though it were a traditional select, there is only a small percentage of chance that it will properly fire the selectItem action.  After some testing I believe that what is happening is that you have the selectItem action set to fire on click, but you also have a focusOut handler set.  The div that holds all of the options is not part of the input that the focusOut handler is set on so when you click an option, it creates a race condition to see which (the focusOut or the click) event will fire first.  I found that if you increase the timeout on your focus handler to even 200 from 100, the click events fire properly 100% of the time.  At least from my initial testing.

**Comment by toranb** [comment]
If a solid 200 works what downside do we get with making the change?

Also -doing a little research ... 200 appears to be the magic number as you suggested :)

https://github.com/klclee/ember-cli-suggest/blob/master/addon/mixins/suggest.js#L42

**Comment by toranb** [comment]
I just published 0.1.0 (and it includes the new timeout of 200 as suggested). Please give it a look and report back if you find it's still half baked :)

http://emberjs.jsbin.com/domuvo/1

I also fixed the non legit jsbin in the readme to what you see above
